feat(sqllogictest): use serde derived structs for schedule parsing

AndreaBozzo opened the pull request:
This PR refactors the schedule file parsing in the sqllogictest crate to use serde-derived structs instead of manual TOML parsing, as requested in #1952.

### Changes

- **New structs with serde derives:**
  - `ScheduleConfig` - top-level configuration parsed from TOML
  - `EngineConfig` - per-engine configuration with `#[serde(flatten)]` for extensibility
  - `EngineType` - enum of supported engine types

- **Refactored parsing flow:**
  - `Schedule::from_file()` now uses `toml::from_str()` directly
  - Added `instantiate_engines()` to separate parsing from engine creation
  - Removed manual `parse_engines()` and `parse_steps()` functions

- **Forward-compatibility:**
  - Uses `#[serde(flatten)]` to capture extra fields in `EngineConfig.extra`
  - This enables PR #1943 to easily add `catalog_type` and `catalog_properties` support

### Relation to #1943

This PR was suggested by @liurenjie1024 as a prerequisite to #1943 (dynamic catalog configuration). The `#[serde(flatten)]` approach allows #1943 to simply extract the catalog configuration from `EngineConfig.extra` without modifying the parsing logic.

### Testing

- All existing tests pass
- Added new unit tests for deserialization behavior
- Integration test with `df_test.toml` passes unchanged

Closes #1952

liurenjie1024 commented on crates/sqllogictest/src/engine/mod.rs:
```diff
@@ -17,29 +17,44 @@
+
+/// Configuration for a single engine instance
+#[derive(Debug, Clone, Deserialize)]
+pub struct EngineConfig {
```
```suggestion
pub enum EngineConfig {
  DataFusion {
     catalog: DatafusionCatalogConfig
  }
}

pub struct DatafusionCatalogConfig {
  typ: String,
  properties: HashMap<String, String> 
}

```

liurenjie1024 commented on crates/sqllogictest/src/engine/mod.rs:
```diff
@@ -17,29 +17,44 @@
+    /// Additional configuration fields for extensibility
+    /// This allows forward-compatibility with future fields like catalog_type, catalog_properties
+    #[serde(flatten)]
+    pub extra: HashMap<String, toml::Value>,
```
+1, see the suggestion above.

liurenjie1024 commented on crates/sqllogictest/src/schedule.rs:
```diff
@@ -105,103 +125,116 @@ impl Schedule {
+        // Test forward-compatibility with extra fields (for PR #1943)
         let input = r#"
+            [engines]
+            df = { type = "datafusion", catalog_type = "rest", some_future_field = "value" }
```
```suggestion
            df = { type = "datafusion", 
            catalog = {
               type = "rest",
               props = {
                  // A hashmap of properties.
                  url = "http://localhost:9000",
                  use_x = "true",
               }
              }
            }       
```
This is a complete example, and then we could use catalog loader to load a catalog.

AndreaBozzo commented:
Thanks for the detailed feedback @liurenjie1024 and @CTTY! I appreciate a lot your time and patience on this.

I've addressed the review comments in commit [f09e5e7](https://github.com/apache/iceberg-rust/commit/f09e5e77c387b5ea309ffd61ab809e94a9b67307), the previous one lacked fmt

The memory catalog remains the default for now. Full catalog loader integration (REST, Glue, HMS, etc.) would have added MANY new dependencies and i wanted to keep it simple to review/modify eventually.

Thanks again.

liurenjie1024 commented on crates/sqllogictest/src/engine/mod.rs:
```diff
@@ -17,29 +17,45 @@
-const TYPE_DATAFUSION: &str = "datafusion";
+/// Configuration for the catalog used by an engine
+#[derive(Debug, Clone, Deserialize)]
+pub struct CatalogConfig {
```
```suggestion
pub struct DatafusionCatalogConfig {
```

We will support more engines in future, for example spark engine, and this catalog config is datafusion specific.
